### ISSUE TITLE
spacyr performance enhancement by combining texts

### ISSUE THREAD

**Issue by amatsuo** [quanteda/spacyr#90]
@kbenoit 
As I indicated, I started working on improving the speed of `spacyr_parse()` one thing I have tried is to provide an option to exclude `ner` (named entity recognition) in `spacy_initialize()`. If the user does not need `entity` this will speed up by about a third (I can put benchmark results if requested).   

The next thing I am thinking to do is to reduce the number of NLP by combining a fixed number of texts and process together. I haven't implemented the actual processing but have tested what is the performance gain if we combine the text. The data I used was the sentence level corpus of Presidential Inaugural Speeches. 

Code:
```{r}
library(spacyr)
library(quanteda)
library(dplyr)
library(microbenchmark)
spacy_initialize()

## function definition
data_text_inaugural_sentence <- data_corpus_inaugural %>% 
    corpus_reshape("sentences") %>% as.character()

spacy_parse_chunk <- function(x) {
    split(data_text_inaugural_sentence, ceiling(seq_along(data_text_inaugural_sentence)/x)) %>%
        lapply(function(y) paste(y, collapse = " ")) %>% unlist() %>% 
        spacy_parse()
}

## test
n <- 10
microbenchmark(spacy_parse_chunk(1),
               spacy_parse_chunk(2),
               spacy_parse_chunk(5),
               spacy_parse_chunk(10),
               spacy_parse_chunk(20),
               spacy_parse_chunk(50),
               spacy_parse_chunk(100),
               spacy_parse_chunk(200),
               spacy_parse_chunk(500),
               spacy_parse_chunk(1000), times = n)
```

Output (using my office desktop):
```
Unit: seconds
                    expr      min       lq     mean   median       uq      max
    spacy_parse_chunk(1) 66.16426 66.31356 66.74909 66.52520 66.98660 67.94092
    spacy_parse_chunk(2) 44.11844 44.39539 44.52556 44.45780 44.59506 45.16326
    spacy_parse_chunk(5) 32.51518 32.59832 32.68386 32.66644 32.71665 32.98154
   spacy_parse_chunk(10) 28.69219 28.98529 29.10786 29.07667 29.16465 29.68750
   spacy_parse_chunk(20) 26.40571 26.57262 26.78336 26.81384 26.98722 27.16832
   spacy_parse_chunk(50) 25.27421 25.42956 26.15504 25.97557 26.63758 27.77997
  spacy_parse_chunk(100) 25.04468 25.25904 25.59844 25.61317 25.71812 26.68067
  spacy_parse_chunk(200) 25.61965 25.82674 26.04774 25.95238 26.33137 26.58659
  spacy_parse_chunk(500) 26.43901 26.60324 27.15675 26.74060 27.35458 28.92657
 spacy_parse_chunk(1000) 26.47361 26.68461 26.83848 26.79811 27.05855 27.15286
 neval
    10
    10
    10
    10
    10
    10
    10
    10
    10
    10
```
The original text vector has 5000 elements. It seems that up to some point it will speed up. 

What do you think? Is this worth pursuing?

**Comment by kbenoit** [comment]
Excellent idea, but there are two ideas here: a) turning off some functionality when you initialise (e.g. NER), and b) improving performance by sending different numbers of documents of different lengths to spaCy. Let's move the first to a new issue, and focus here on the second.

In the [code for corpus.zip](https://github.com/quanteda/quanteda/blob/master/R/corpuszip.R) I defined a global to be a marker between texts, which can be used to [combine texts](https://github.com/quanteda/quanteda/blob/master/R/corpus.R#L212-L217) or [extract texts](https://github.com/quanteda/quanteda/blob/master/R/corpuszip.R#L6-L9) from the compressed text vector, which has to be a single character object for `memCompress()`.

In this case it looks like you want to do the opposite: split documents but while ensuring that you do not split sentences, and then recombine them. This is easy using `corpus_reshape()` but splitting everything into sentences may also be inefficient. Probably you will want to run some length v. number of document curves to find the sweet spot.

**Comment by amatsuo** [comment]
@kbenoit 
I have implemented this by inserting a `random_string + docname + .` as a head of each text and then concatenate. It works fine, but I found some issues that is if we combine text, the result of parsing changes. Here is a example:

```r
library(spacyr)
spacy_initialize()
data_text_inaugural_sentence <- quanteda::data_corpus_inaugural %>% 
    corpus_reshape("sentences")


tmp <- spacy_parse(data_text_inaugural_sentence)
dim(tmp)
tmp2 <- spacy_parse(data_text_inaugural_sentence, batch_size = 100)
dim(tmp2)
equiv <- apply(tmp == tmp2, 1, function(x) sum(x == F))
table(equiv)
## equiv
##      0      1      2      3 
## 148002    849    749      3 

apply(tmp == tmp2, 2, function(x) sum(x == F))
##      doc_id sentence_id    token_id       token       lemma         pos      entity 
##           0         820         842           0           5          88         601 
```
The additional texts seem to alter the results of parcing. I've tried several other possible sentences but it didn't work better. Probably, we should give up the idea, but the speed is a bit attractive:

```
> microbenchmark::microbenchmark(
+     no_batch = spacy_parse(data_text_inaugural_sentence),
+     batch = spacy_parse(data_text_inaugural_sentence, batch_size = 100),
+     times = 1)
Unit: seconds
     expr      min       lq     mean   median       uq      max neval
 no_batch 85.72393 85.72393 85.72393 85.72393 85.72393 85.72393     1
    batch 36.23729 36.23729 36.23729 36.23729 36.23729 36.23729     1
```
You can try this by installing the latest `performance-enhancement` branch.

**Comment by methodds** [comment]
Just out of curiosity: Did you also run comparisons between doing everything in python and using `spacyr`? It would be interesting to see how costly the R wrapper is. I just parsed 8000 short documents and it took quite a while.

**Comment by kbenoit** [comment]
While this may not connect perfectly to the above, I ran some tests after finding the parsing of the SOTU corpus to be really slow. But I put the code below since it is easily adapted for any corpus of numerous documents. I didn't have the patience to parse all 232 SOTU addresses, but that is something that you @amatsuo should try. 

Basic findings:

1. The timing is not linear with the number of documents, but instead starts to increase.  
2. Numerous smaller documents are faster to parse than their concatenated equivalents, after about 35 documents. Fewer than 35 SOTU speeches and the performance is the same.

```r

library("spacyr")
packageVersion("spacyr")
# [1] ‘0.9.6’
library("quanteda")
# Package version: 1.1.2
# Parallel computing: 2 of 8 threads used.

data(data_corpus_sotu, package = "quanteda.corpora")

spacy_initialize()
# Finding a python executable with spacy installed...
# spaCy (language model: en) is installed in more than one python
# spacyr will use /anaconda/bin/python (because ask = FALSE)
# successfully initialized (spaCy Version: 2.0.9, language model: en)


ndocuments <- c(1, seq(5, 50, 5)) # ndoc(data_corpus_sotu), 5))) {

cat("Parse times:\n")
timing <- c()
for (i in ndocuments) { 
    cat(" ", i, ": ", sep = "")
    thistiming <- system.time(
        spacy_parse(texts(data_corpus_sotu)[seq_len(i)],
                    tag = FALSE, lemma = FALSE, entity = FALSE, dependency = FALSE),
        gcFirst = TRUE
    )
    timing <- c(timing, thistiming["elapsed"])
    cat(tail(timing, 1), "\n")
}
# 1: 0.229 
# 5: 1.732 
# 10: 3.568 
# 15: 5.242 
# 20: 8.21 
# 25: 10.812 
# 30: 14.928 
# 35: 21.907 
# 40: 30.862 
# 45: 38.432 
# 50: 45.638 


# parse all at once
timing2 <- c()
for (i in ndocuments) { 
    cat(" ", i, ": ", sep = "")
    thistiming <- system.time(
        # this usage of texts() concatenates all texts into one
        spacy_parse(texts(data_corpus_sotu[seq_len(i)], groups = rep(1, i)),
                    tag = FALSE, lemma = FALSE, entity = FALSE, dependency = FALSE),
        gcFirst = TRUE
    )
    timing2 <- c(timing2, thistiming["elapsed"])
    cat(tail(timing2, 1), "\n")
}
# 1: 0.239 
# 5: 1.519 
# 10: 3.823 
# 15: 5.634 
# 20: 8.077 
# 25: 11.161 
# 30: 14.728 
# 35: 22.687 
# 40: 35.806 
# 45: 46.765 
# 50: 86.413 

library("ggplot2")
df <- data.frame(ndocuments = rep(ndocuments, times = 2),
                 elapsed = c(timing, timing2),
                 method = c(rep("Separate documents", length(ndocuments)),
                            rep("Concatenated", length(ndocuments)))
                 )
ggplot(df, aes(x = ndocuments, y = elapsed)) +
    geom_line(aes(color = method)) + 
    geom_point(aes(color = method))    
```
![rplot](https://user-images.githubusercontent.com/2182246/37492692-3ac3abae-289a-11e8-8b71-98fc3da073cf.png)

**Comment by amatsuo** [comment]
@kbenoit 

I implemented a simple pipe method. That was much simpler to do than I thought...

It seems working fine and performance is better. 

```r
```

**Comment by amatsuo** [comment]
This was mostly resolved in #107. 

Text concatenation sounds a good idea but there seems to be no way to make spacy behave exactly the same for concatenated texts and individual tests (before concatenation). Probably we should prioritize the reliability of results over the marginal efficiency gain.
